### ISSUE TITLE
steal-electron IPC require issue

### ISSUE THREAD

**Issue by justinbmeyer** [stealjs/stealjs#42]
@gKreator commented on [Mon Apr 17 2017](https://github.com/stealjs/steal-electron/issues/20)

This is most likely my own derpiness rather than a code issue.

https://www.christianengvall.se/ipcmain-and-ipcrenderer/
According to this tutorial, and a few others, to get IPC working so that I can communicate from the 'main' process to the 'renderer' process. I need to include electron on the render process.

This works in the console.
const {ipcRenderer} = require('electron'); 

If I include this via stealjs es6
import { ipcRenderer } from "electron";

I get errors like this... it wants me to require path and fs js which are both node libraries and not browser based I don't think and I get errors like this.
![image](https://cloud.githubusercontent.com/assets/1431487/25108059/c7f64214-2398-11e7-843d-91ec046a6809.png)

If I fix those problems I just get other issues and so forth.

Thanks 

---

@justinbmeyer commented on [Mon Apr 17 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294636286)

You need this code to use the node `require` method, not `steal`'s `require`.  You can get it by importing `@node-less` like this:

https://github.com/stealjs/steal-less/blob/master/less-engine-node.js

---

@gKreator commented on [Mon Apr 17 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294655004)

I tried:
`import nodeRequire from "@node-require";`

![image](https://cloud.githubusercontent.com/assets/1431487/25112073/77367cce-23b4-11e7-8c05-e509c4a19873.png)

It does build, but it crashes.
Error in console:
![image](https://cloud.githubusercontent.com/assets/1431487/25112082/8141f3d8-23b4-11e7-8dde-3dc7a55ba1b5.png)

---

@justinbmeyer commented on [Mon Apr 17 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294661166)

Hmmm, it seems that node-require doesn't exist in this realm.  This might be due to steal not understanding the environment it's in.  

Look for the following code in your version of `steal.js`: https://github.com/stealjs/steal/blob/4ebf89f8d85b0542fddf7143a0ec7e145e06c521/src/node-require.js#L2 

Is `isNode` true?  

If not look for the following code in `steal.js`: https://github.com/stealjs/steal/blob/7422b98b8d4a73e50c5f4f47fb8f169d44d797cd/src/start.js#L115

What is failing w/ the environment detection?

---

@justinbmeyer commented on [Mon Apr 17 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294661328)

ah, the electron detection will change `isNode`: https://github.com/stealjs/steal/blob/7422b98b8d4a73e50c5f4f47fb8f169d44d797cd/src/start.js#L125

Try removing that.  Does everything work?  I'm sure @matthewp can help make sense of that tomorrow.  

---

@matthewp commented on [Tue Apr 18 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294848902)

`@node-require` doesn't exist in steal.js, only in the Node main for steal. It depends on running within node_modules/steal/ which won't be the case when you do your build.  I think you want to use `System._nodeRequire` instead.  You can get it best with:

```js
var nodeRequire = require("@loader")._nodeRequire;

nodeRequire("whatever");
```

However I haven't tested this with Electron so there might be a bug there. File an issue in stealjs/steal if that doesn't work.

---

@garysydcon commented on [Tue Apr 18 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294874959)

I am at work right now so I cannot test anything till later, but I am just wondering how to do `var nodeRequire = require("@loader")._nodeRequire;` in ES6. Would it be `import {_nodeRequire as nodeRequire} from "@loader"`?

---

@matthewp commented on [Tue Apr 18 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-294899805)

No, it would be:

```js
import loader from "@loader";
const nodeRequire = loader._nodeRequire;
```

---

@gKreator commented on [Tue Apr 18 2017](https://github.com/stealjs/steal-electron/issues/20#issuecomment-295021620)

Thank you both, it works!!! 
Please maybe add this in the README for the next guy!

**Comment by matthewp** [comment]
I don't think we should document it, people will have false expectations for how it works. They will expect it to be a local Node.js `require` function for the module they are using it in. So they would expect `nodeRequire('./foo.js')` to load that file relative to the current module. Since code executed within Steal is not part of Node's module system, there *is no* local Node require we can give to a user.

`@node-require` attempts to fix this, with the caveat that it has Node dependencies (and therefore expects to be run within node_modules/steal). With npm3+ it's likely that it will still work for this use-case anyways. Maybe better to document that.

**Comment by gKreator** [comment]
Yes, I understand the issues here. I would suggest to just mentioning it on the steal-electron README or link to the issue, because using electron like this is very common.
